File: client/android/app/src/main/java/com/sonfy/app/SonfyWebView.java
package com.sonfy.app;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.webkit.CookieManager;
import android.webkit.WebSettings;
import android.webkit.WebView;

/**
 * Custom WebView that keeps playing audio in background.
 * The key trick is overriding onWindowVisibilityChanged to always report VISIBLE,
 * which prevents Android from pausing the WebView's media playback.
 */
public class SonfyWebView extends WebView {

    public SonfyWebView(Context context) {
        super(context);
        init();
    }

    public SonfyWebView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public SonfyWebView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        WebSettings settings = getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);
        settings.setMediaPlaybackRequiresUserGesture(false);
        settings.setSupportZoom(true);
        settings.setBuiltInZoomControls(true);
        settings.setDisplayZoomControls(false);
        settings.setDatabaseEnabled(true);
        settings.setAllowFileAccess(true);
        settings.setAllowContentAccess(true);
        
        // Enable cookies
        CookieManager.getInstance().setAcceptCookie(true);
        CookieManager.getInstance().setAcceptThirdPartyCookies(this, true);
        
        // Enable focus for keyboard input
        setFocusable(true);
        setFocusableInTouchMode(true);
        
        // Hardware acceleration for better performance
        setLayerType(View.LAYER_TYPE_HARDWARE, null);
    }

    /**
     * CRITICAL: Override this method to always report VISIBLE.
     * This prevents Android from pausing the WebView's media playback
     * when the app goes to background.
     */
    @Override
    protected void onWindowVisibilityChanged(int visibility) {
        // Always call super with VISIBLE to keep WebView active
        super.onWindowVisibilityChanged(View.VISIBLE);
    }
}

File: app/src/main/java/com/sonfy/app/SonfyWebView.kt
package com.sonfy.app

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.webkit.CookieManager
import android.webkit.WebSettings
import android.webkit.WebView

/**
 * Custom WebView that keeps playing audio in background.
 * 
 * The key trick is overriding onWindowVisibilityChanged to always report VISIBLE,
 * which prevents Android from pausing the WebView's media playback when the app
 * goes to background.
 * 
 * This is the same approach used by NouTube.
 */
class SonfyWebView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : WebView(context, attrs, defStyleAttr) {

    init {
        settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
            mediaPlaybackRequiresUserGesture = false
            databaseEnabled = true
            allowFileAccess = true
            allowContentAccess = true
            @Suppress("DEPRECATION")
            allowUniversalAccessFromFileURLs = true
            @Suppress("DEPRECATION")
            allowFileAccessFromFileURLs = true
            setSupportZoom(true)
            builtInZoomControls = true
            displayZoomControls = false
            loadWithOverviewMode = true
            useWideViewPort = true
            mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        }
        
        // Enable cookies
        CookieManager.getInstance().apply {
            setAcceptCookie(true)
            setAcceptThirdPartyCookies(this@SonfyWebView, true)
        }
        
        // Enable focus for keyboard input
        isFocusable = true
        isFocusableInTouchMode = true
        
        // Hardware acceleration
        setLayerType(View.LAYER_TYPE_HARDWARE, null)
    }

    /**
     * CRITICAL: Override this method to always report VISIBLE.
     * 
     * When Android detects that the app is in background, it calls this method
     * with INVISIBLE or GONE. By always calling super with VISIBLE, we trick
     * the WebView into thinking it's always visible, preventing it from pausing
     * media playback.
     */
    override fun onWindowVisibilityChanged(visibility: Int) {
        // Always report as VISIBLE to keep media playing in background
        super.onWindowVisibilityChanged(View.VISIBLE)
    }
}
